mantidproject/mantid
# Add DepolarizedAnalyserTransmission Algorithm

Part of the Polarised SANS Epic. 

### Description of work

#### Summary of work
Adds an algorithm which determines the transmission of a depolarised He3 cell from a depolarised run and an empty cell run. 

Fixes #36383 

#### Further detail of work
Normalises the depolarised workspace by the empty cell workspace and then performs an exponential fit to determine the T_E (empty cell transmission) and pxd (gas depth * pressure). This can then be used in later algorithms to determine the polarisation efficiency for a given wavelength. 

Both input workspaces need to contain a single spectrum and be in wavelength. 

### To test:
1. This basic example from the documentation will provide a decent example of the algorithm in action:
```
CreateSampleWorkspace(OutputWorkspace='mt', Function='User Defined', UserDefinedFunction='name=UserFunction, Formula=1.465e-07*exp(0.0733*4.76*x)', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=3.5, XMax=16.5, BinWidth=0.1)
CreateSampleWorkspace(OutputWorkspace='dep', Function='User Defined', UserDefinedFunction='name=UserFunction, Formula=0.0121*exp(-0.0733*10.226*x)', XUnit='wavelength', NumBanks=1, BankPixelWidth=1, XMin=3.5, XMax=16.5, BinWidth=0.1)

output = SANSCalcDepolarisedAnalyserTransmission("dep", "mt")

print("PXD Value = " + str(output.column("Value")[0]) + ".")
print("T_E Value = " + str(output.column("Value")[1]) + ".")
```
Note: If you want to test with real data to test with, message me for some. 
2. Check the documentation is formatted properly and can be accessed from the `?` in the algorithm dialog. (You may need to build the `docs-qthelp` target for this to work. 
3. Try the algorithm with some *terrible* fit starting params. It should fail with a runtime error. (Note: The algorithm is designed to fail fast and throw if something goes wrong rather than warning. This is often done near the beginning of a polarisation correction, so it being wrong would throw off a lot).
4. Try the algorithm with an all 0 workspace as the `mt` file. This should also fail with a helpful error message. 

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
